stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -666,24 +676,71 @@ extension Unsafe${Mutable}BufferPointer {
+  ///     buffer. The type `T` must be layout compatible
   ///     with the pointer's `Element` type.
   ///   - body: A closure that takes a ${Mutable.lower()} typed buffer to the
-  ///     same memory as this buffer, only bound to type `T`. The buffer argument 
-  ///     contains the same number of complete instances of `T` as the original  
-  ///     buffer’s `count`. The closure's buffer argument is valid only for the 
-  ///     duration of the closure's execution. If `body` has a return value, that 
-  ///     value is also used as the return value for the `withMemoryRebound(to:_:)` 
+  ///     same memory as this buffer, only bound to type `T`. The buffer
+  ///     parameter contains a number of complete instances of `T` based
+  ///     on the capacity of the original buffer and the stride of `Element`.
+  ///     The closure's buffer argument is valid only for the duration of the
+  ///     closure's execution. If `body` has a return value, that value
+  ///     is also used as the return value for the `withMemoryRebound(to:_:)`
   ///     method.
+  ///   - buffer: The buffer temporarily bound to `T`.
   /// - Returns: The return value, if any, of the `body` closure parameter.
   @inlinable // unsafe-performance
+  @_alwaysEmitIntoClient
+% if Mutable:
+  @_silgen_name("$_swift_UnsafeMutableBufferPointer_se0333_withMemoryRebound")
+% else:
+  @_silgen_name("$_swift_UnsafeBufferPointer_se0333_withMemoryRebound")
+% end
   public func withMemoryRebound<T, Result>(
+    to type: T.Type,
+    _ body: (_ buffer: ${Self}<T>) throws -> Result
+  ) rethrows -> Result {
+    guard let base = _position?._rawValue else {
+      return try body(.init(start: nil, count: 0))
+    }
+
+    let newCount: Int
+    if MemoryLayout<T>.stride == MemoryLayout<Element>.stride {
+      newCount = count
+      _debugPrecondition(
+        MemoryLayout<T>.alignment == MemoryLayout<Element>.alignment
+      )
+    } else {
+      newCount = count * MemoryLayout<Element>.stride / MemoryLayout<T>.stride
+      _debugPrecondition(
+        Int(bitPattern: .init(base)) & (MemoryLayout<T>.alignment-1) == 0 &&
+        MemoryLayout<T>.stride > MemoryLayout<Element>.stride
+        ? MemoryLayout<T>.stride % MemoryLayout<Element>.stride == 0
+        : MemoryLayout<Element>.stride % MemoryLayout<T>.stride == 0
+      )
+    }
+    let binding = Builtin.bindMemory(base, newCount._builtinWordValue, T.self)
+    defer { Builtin.rebindMemory(base, binding) }
+    return try body(.init(start: .init(base), count: newCount))
+  }
+
+  @available(*, unavailable)
+% if Mutable:
+  @_silgen_name("$sSr17withMemoryRebound2to_qd_0_qd__m_qd_0_Sryqd__GKXEtKr0_lF")
+% else:
+  @_silgen_name("$sSR17withMemoryRebound2to_qd_0_qd__m_qd_0_SRyqd__GKXEtKr0_lF")
+% end
+  public func _legacy_withMemoryRebound<T, Result>(
     to type: T.Type, _ body: (${Self}<T>) throws -> Result
   ) rethrows -> Result {
     if let base = _position {
```

Comment:
This preserves the old implementation verbatim, but we could also forward to the new implementation. This would be better from a _testability_ standpoint, but less rigourously correct from a compatibility standpoint. However, we are lifting a restriction so the new implementation should be compatible. Thoughts?